### PR TITLE
Add support for noopener and noreferrer in anchors

### LINKED ISSUE #89: Add rel="noopener" to anchor if target="_blank" is set => security enhancement
**Add rel="noopener" to anker if target="_blank" is set**
Based on the OWASP article
https://owasp.org/www-community/attacks/Reverse_Tabnabbing
it would be nice if the noopener attribute would be set automatically if the target blank attribute is in use.

This is very similar to the nofollow setting in antisamy

**Example**
`<a href="https://example.com" target="_blank"> => <a href="https://example.com" target="_blank" rel="noopener">`

### PR DESCRIPTION
**Pull request by spassarop**
Fixes #89. This adds a directive (false by default) to allow adding `noopener` and `noreferrer` (at the same time) on anchor tags, just like `nofollow` but work separately.

With DOM, attributes are processed and AntiSamy then adds or not the corresponding values to `rel` attribute.
With SAX, AntiSamy adds or not the corresponding values to `rel` attribute depending on directive values and in that moment checks valid values for the `rel` attribute. It's more like a backward compatibility behavior shown in a `nofollow` test case.

Default policy has the new directive and allows the presence of `nofollow`, `noopener` and `noreferrer` only (only `nofollow` before) with a regular expression.

### CODE REVIEW

**Comment by spassarop** [comment]
@davewichers well I forgot the logger, I'm on it. 

One thing I did not what to do with was the placing of the static method with the logic for adding or not the `rel` values. `Attribute` class was the only place I found suitable.

**Comment by spassarop** [comment]
Hope those logging additions are what you were expecting. I am not really used to manage logging in Java (I just use Java with AntiSamy) so maybe my approach was not the best.

File: src/main/java/org/owasp/validator/html/Policy.java [hunk 1]
```diff
@@ -338,6 +339,8 @@ protected Policy(ParseContext parseContext) {
         this.directives = Collections.unmodifiableMap(parseContext.directives);
         this.globalAttributes = Collections.unmodifiableMap(parseContext.globalAttributes);
         this.dynamicAttributes = Collections.unmodifiableMap(parseContext.dynamicAttributes);
+
```
Comment (davewichers) [inline src/main/java/org/owasp/validator/html/Policy.java:342]: Can you remove unnecessary changes like this?

File: src/main/java/org/owasp/validator/html/model/Attribute.java [hunk 2]
```diff
@@ -143,6 +143,35 @@ public String matcherRegEx(boolean hasNext){
             }
         }
         return regExp.toString();
+    }
+
+    public static String mergeRelValuesInAnchor(boolean addNofollow, boolean addNoopenerAndNoreferrer, String currentRelValue) {
```
Comment (davewichers) [inline src/main/java/org/owasp/validator/html/model/Attribute.java:148]: Please add JavaDoc explaining this method signature.

File: src/main/java/org/owasp/validator/html/model/Attribute.java [hunk 3]
```diff
@@ -143,6 +143,35 @@ public String matcherRegEx(boolean hasNext){
...
+        String newRelValue = "";
+        if (currentRelValue == null || currentRelValue.isEmpty()) {
+            if (addNofollow) newRelValue = "nofollow";
+            if (addNoopenerAndNoreferrer) newRelValue += " noopener noreferrer";
+        }
```
Comment (davewichers) [inline src/main/java/org/owasp/validator/html/model/Attribute.java:153]: Combine these two lines into one line:  } else {

File: src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java [hunk 4]
```diff
@@ -368,8 +372,26 @@ private void actionValidate(int currentStackDepth, Element ele, Node parentNode,
...
+        if ("a".equals(tagNameLowerCase)) {
+            boolean addNofollow = isNofollowAnchors;
+            boolean addNoopenerAndNoreferrer = false;
+
+            if (isNoopenerAndNoreferrerAnchors) {
+                Node targetAttribute = ele.getAttributes().getNamedItem("target");
+                if (targetAttribute != null && targetAttribute.getNodeValue().equalsIgnoreCase("_blank")) {
+                    addNoopenerAndNoreferrer = true;
+                }
+            }
```
Comment (davewichers) [inline src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java:384]: Combine these two lines too.

File: src/main/java/org/owasp/validator/html/scan/MagicSAXFilter.java [hunk 5]
```diff
@@ -66,20 +68,24 @@
...
+	private static final Logger logger = LoggerFactory.getLogger(MagicSAXFilter.class);
+
     public MagicSAXFilter(ResourceBundle messages) {
 		this.messages = messages;
     }
 
     public void reset(InternalPolicy instance){
         this.policy = instance;
         isNofollowAnchors = policy.isNofollowAnchors();
+		isNoopenerAndNoreferrerAnchors = policy.isNoopenerAndNoreferrerAnchors();
```
Comment (davewichers) [inline src/main/java/org/owasp/validator/html/scan/MagicSAXFilter.java:88]: This line has a different type of white space in front. Please change to match the previous/following lines.

File: src/main/java/org/owasp/validator/html/scan/MagicSAXFilter.java [hunk 6]
```diff
@@ -363,8 +369,33 @@ public void startElement(QName element, XMLAttributes attributes, Augmentations
...
+					if ("a".equals(element.localpart)) {
+						boolean addNofollow = isNofollowAnchors;
+						boolean addNoopenerAndNoreferrer = false;
+
+						if (isNoopenerAndNoreferrerAnchors) {
+							String targetValue = attributes.getValue("target");
+							if (targetValue != null && targetValue.equalsIgnoreCase("_blank")) {
+								addNoopenerAndNoreferrer = true;
+							}
+						}
```
Comment (davewichers) [inline src/main/java/org/owasp/validator/html/scan/MagicSAXFilter.java:381]: Combine these 2 lines.

File: src/main/resources/antisamy.xml [hunk 7]
```diff
@@ -795,9 +796,9 @@ http://www.w3.org/TR/html401/struct/global.html
 				</regexp-list>
 			</attribute>
 			<attribute name="rel">
-				<literal-list>
-					<literal value="nofollow"/>
-				</literal-list>
+				<regexp-list>
+					<regexp value="(\s*(nofollow|noopener|noreferrer)(\s(nofollow|noopener|noreferrer))*\s*)?"/>
```
Comment (davewichers) [inline src/main/resources/antisamy.xml:800]: Does this regex allow duplicate entries? If so, it there a way to create a regex that allows these 3, but no duplicates of them?

File: src/main/java/org/owasp/validator/html/model/Attribute.java (hunk 3 above)
Comment (spassarop) [reply src/main/java/org/owasp/validator/html/model/Attribute.java:153]: There is no else. The two cases are independent and can happen both or just one.

File: src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java (hunk 4 above)
Comment (spassarop) [reply src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java:384]: Do you mean in a single if with `&&`? That would repeat `ele.getAttributes().getNamedItem("target")` for the `null` check and the actual use, seems more readable this way, unless you mean other thing (or disagree).

File: src/main/java/org/owasp/validator/html/scan/MagicSAXFilter.java (hunk 6 above)
Comment (spassarop) [reply src/main/java/org/owasp/validator/html/scan/MagicSAXFilter.java:381]: Same as the `ele.getAttributes().getNamedItem("target")` when we reach agreement I do the same resolution.

File: src/main/resources/antisamy.xml (hunk 7 above)
Comment (spassarop) [reply src/main/resources/antisamy.xml:800]: It does allow duplicates. The only way to allow the three words with no duplicates regardless the order is to specify a regex with all possible combinations which would look long and ugly. That's why I left it to at least have the valid words with compulsory spaces (if there are more than one).

File: src/main/java/org/owasp/validator/html/model/Attribute.java (hunk 2 above)
Comment (spassarop) [reply src/main/java/org/owasp/validator/html/model/Attribute.java:159]: Done

File: src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java (hunk 4 above)
Comment (davewichers) [reply src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java:380]: > Do you mean in a single if with `&&`? That would repeat `ele.getAttributes().getNamedItem("target")` for the `null` check and the actual use, seems more readable this way, unless you mean other thing (or disagree).

No - all of these combine line comments refer to having the else line be:   } else {    (all on one line). Simply a formatting comment, not code change.

Comment (spassarop) [reply src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java:380]: Oh I get it now. This will be the only `else` left then.

Comment (spassarop) [reply src/main/java/org/owasp/validator/html/scan/AntiSamyDOMScanner.java:380]: And it's gone now
